File: Medium/2976. Minimum Cost to Convert String I.java
class Solution {
    public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
        final var dists = new int[26][26];
        for (int i = 0; i < 26; i++) {
            for (int j = 0; j < 26; j++) {
                if (i != j) {
                    dists[i][j] = Integer.MAX_VALUE;
                }
            }
        }
        for (int i = 0; i < cost.length; i++) {
            dists[original[i] - 'a'][changed[i] - 'a'] = Math.min(dists[original[i] - 'a'][changed[i] - 'a'], cost[i]);
        }
        for (int k = 0; k < 26; k++) {
            for (int i = 0; i < 26; i++) {
                if (dists[i][k] == Integer.MAX_VALUE) {
                    continue;
                }
                for (int j = 0; j < 26; j++) {
                    if (dists[k][j] == Integer.MAX_VALUE) {
                        continue;
                    }
                    dists[i][j] = Math.min(dists[i][j], dists[i][k] + dists[k][j]);
                }
            }
        }
        var res = 0L;
        for (int i = 0; i < source.length(); i++) {
            int s = source.charAt(i) - 'a';
            int t = target.charAt(i) - 'a';
            if (dists[s][t] == Integer.MAX_VALUE) {
                return -1;
            }
            res += (long) dists[s][t];
        }
        return res;
    }
}

File: Medium/2976. Minimum Cost to Convert String I.py
from cmath import inf
from typing import List


class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        min_dists = [[inf] * 26 for _ in range(26)]
        for i in range(26):
            min_dists[i][i] = 0
        for s, t, c in zip(original, changed, cost):
            u = ord(s) - 97
            v = ord(t) - 97
            min_dists[u][v] = min(min_dists[u][v], c)
        for y in range(26):
            for x in range(26):
                if min_dists[x][y] == inf:
                    continue
                for z in range(26):
                    if min_dists[y][z] == inf:
                        continue
                    min_dists[x][z] = min(min_dists[x][z], min_dists[x][y] + min_dists[y][z])
        total_costs = 0
        for s, t in zip(source, target):
            if s == t:
                continue
            u = ord(s) - 97
            v = ord(t) - 97
            if min_dists[u][v] == inf:
                return -1
            total_costs += min_dists[u][v]
        return total_costs # type: ignore
        
